### ISSUE TITLE
this is damn serious: there is nothing that blocks InsertAllAsync from being executed in parallel with a RunInTransactionAsync

### ISSUE THREAD

**Issue by csm101** [praeclarum/sqlite-net#1190]
If you are running a transaction with RunInTransactionAsync(), other tasks are still free to execute write commands while your transaction is still running and hasn't committed yet its work. 

  For example: let's say you are running a long transaction on your db and another task tries to write  a bunch of records with InsertAllAsync:  if you log what it is happening in the db, you'll see that the "inserts" of the other task get mixed with the work of your transaction.
  **This means that if your transaction fails, all the inserts made by the other task will be considered part of such transaction and will be rolled back.** (just to mention one of the possible problems)

 Actually SqliteAsyncConnection correctly serializes write acces to the database ONLY if all writes are performed by using RunInTransactionAsync. if ALL your source code writes in the database using only RunInTransactionAsync, then you can be sure that all your transactions will be performed in isolation from the others: you'll get exclusive write access to your db and you will be the only one being allowed to write until your code commits or rollbacks (RunInTransacionAsync takes care of doing it at the end of your write operation)...

But: if there is some code in your application that writes using InsertAsync, DeleteAsync, UpdateAsync, etc, you are pretty fucked, because these calls simply don't care about your transaction being executed...

 These calls don't even care about each other's transaction: if you try to launch three "InsertAll" operations in parallel and you log the sql operations being executed, you will notice that the inserts from one task are interleaved with the inserts of the other.

 **WHERE IS THE PROBLEM?**

Let's start from what it is working correctly: RunInTransactionAsync handles the serialization of its calls by internally calling this TransactAsync private method:

```csharp
		Task<T> TransactAsync<T> (Func<SQLiteConnectionWithLock, T> transact)
		{
			return Task.Factory.StartNew (() => {
				var conn = GetConnectionAndTransactionLock (out var transactionLock);
				lock (transactionLock) {
					using (conn.Lock ()) {
						return transact (conn);
					}
				}
			}, CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
		}

```
Notice that here there are two lock objects involved: transactionLock which is the one really responsible to serialize write access to the internal session. and SqliteConnection.Lock() which actually does... Well IT DOES NOTHING if your database is opened with the FullMutex option (in such case it is really implemented with a do nothing FakeLockWrapper instance).. and SqliteOpenFlags.FullMutex is the default behavior. 
So.. with RunInTransactionAsync that other transactionLock (obtained via GetConnectionAndTransactionLock) saves the day.

All other Write operations, instead gain access to the internal session by calling this other private method:

```csharp
		Task<T> WriteAsync<T> (Func<SQLiteConnectionWithLock, T> write)
		{
			return Task.Factory.StartNew (() => {
				var conn = GetConnection ();
				using (conn.Lock ()) {
					return write (conn);
				}
			}, CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
		}
```

And this other method doesn't care at all of the transactionLock used by RunInTransactionAsync: it uses only the "do-nothing" lock.
In fact such WriteAsync method is IDENTICAL to the the ReadAsync method used by read operations (like ToListAsync() and the like):

they are the same:

```csharp
		Task<T> ReadAsync<T> (Func<SQLiteConnectionWithLock, T> read)
		{
			return Task.Factory.StartNew (() => {
				var conn = GetConnection ();
				using (conn.Lock ()) {
					return read (conn);
				}
			}, CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
		}
```

Now: I solved the issue in my code by reimplementing WriteAsync in order to call TransactAsync. Now WriteAsync looks like this in my code and now all write operations are being correctly separated in their own transactions:

```csharp
   Task<T> WriteAsync<T> (Func<SQLiteConnectionWithLock, T> write) => TransactAsync (write);
```

but... Am I missing something?
